scripts/lib/ts-generator.js

```diff
@@ -0,0 +1,437 @@
+        var waiterModelFile = fs.readFileSync(waiterFilePath);
+        var waiterModel = JSON.parse(waiterModelFile);
+        serviceModel.waiters = waiterModel.waiters;
+    }
+
+    return serviceModel;
+};
+
+/**
+ * Determines if a member is required by checking for it in a list.
+ */
+TSGenerator.prototype.checkRequired = function checkRequired(list, member) {
+    if (list.indexOf(member)) {
+        return true;
+    }
+    return false;
+};
+
+/**
+ * Generates whitespace based on the count.
+ */
+TSGenerator.prototype.tabs = function tabs(count) {
+    var code = '';
+    for (var i = 0; i < count; i++) {
+        code += '  ';
+    }
+    return code;
+};
+
+/**
+ * Transforms documentation string to a more readable format.
+ */
+TSGenerator.prototype.transformDocumentation = function transformDocumentation(documentation) {
+    if (!documentation) {
+        return '';
+    }
+    documentation = documentation.replace(/<(?:.|\n)*?>/gm, '');
+    documentation = documentation.replace(/\*\//g, '*');
+    return documentation;
+};
+
+/**
+ * Returns a doc string based on the supplied documentation.
+ * Also tabs the doc string if a count is provided.
+ */
+TSGenerator.prototype.generateDocString = function generateDocString(documentation, tabCount) {
+    tabCount = tabCount || 0;
+    var code = '';
+    code += this.tabs(tabCount) + '/**\n';
+    code += this.tabs(tabCount) + ' * ' + this.transformDocumentation(documentation) + '\n';
+    code += this.tabs(tabCount) + ' */\n';
+    return code;
+};
+
+/**
+ * Returns an array of custom configuration options based on a service identiffier.
+ * Custom configuration options are determined by checking the metadata.json file.
+ */
+TSGenerator.prototype.generateCustomConfigFromMetadata = function generateCustomConfigFromMetadata(serviceIdentifier) {
+    // some services have additional configuration options that are defined in the metadata.json file
+    // i.e. dualstackAvailable = useDualstack
+    // create reference to custom options
+    var customConfigurations = [];
+    var serviceMetadata = this.metadata[serviceIdentifier];
+    // loop through metadata members
+    for (var memberName in serviceMetadata) {
+        if (!serviceMetadata.hasOwnProperty(memberName)) {
+            continue;
+        }
+        var memberValue = serviceMetadata[memberName];
+        // check configs
+        switch (memberName) {
+            case 'dualstackAvailable':
+                customConfigurations.push(CUSTOM_CONFIG_ENUMS.DUALSTACK);
+                break;
+        }
+    }
+
+    return customConfigurations;
+};
+
+/**
+ * Generates a type or interface based on the shape.
+ */
+TSGenerator.prototype.generateTypingsFromShape = function generateTypingsFromShape(shapeKey, shape, tabCount) {
+    // some shapes shouldn't be generated if they are javascript primitives
+    if (['string', 'boolean', 'number', 'Date', 'Blob'].indexOf(shapeKey) >= 0) {
+        return '';
```

Comment:
Instead of skipping shapes that happen to be named these, we could just add an underscore in front to differentiate the name of the shape from the javascript primitives.
It could still be useful to generate a typing for these shapes. For example, some services (like CodeCommit) have a shape named `"Date"` that is modeled as a string rather than a timestamp. If we rename the shape rather than skip it and have it be typed as a JavaScript `Date` object, then it could prevent errors before runtime, such as calling a `Date` method like `getMonth()` on the string.